YAML content (#9)
mine-cetinkaya-rundel/quarto-tdg: yaml.qmd

In this chapter ...

Options for Quarto projects, documents and code cells are set using a language called [YAML](https://yaml.org).
YAML describes the syntax of setting options, but the options that are available and the values they can take are defined by Quarto.

By now you've seen Quarto options set using YAML syntax in a number of places:

-   in the project configuration file `_quarto.yml`
-   in the document header, i.e. between the `---` at the top of a Quarto document
-   in code cells, on lines prefaced by `#|`

You'll also learn about one new place you can set options, `_metadata.yml`.

## General YAML syntax

In this section you'll learn the necessary YAML fundamentals to avoid syntax errors when setting options.
Although the syntax applies generally to any YAML, all the examples you'll see are valid Quarto options and values.

### YAML Mappings

A YAML mapping is the specification of a key-value pair using the syntax `key: value`.
Values can be simple, like a string, boolean, or number:

``` {.yaml filename="document.qmd"}
---
title: My Document  # A string
toc: true           # A boolean, `true`/`false`

> toc-depth: 2        # A number

Reviewer: Should we use code annotation for these instead of comments? Is that even possible?

Author: Yeah, that might be nice. I'll do it here. But, when I repeat this example later, I'm going to use comments, so I can still have a tip about what YAML comments look like 😆 .

---
```

::: callout-tip
## YAML comments

The hash symbol (`#`) indicates the start of a comment, which is ended by the end of the line.
:::

Or, values can be more complicated like another mapping.
When the value of a key is another mapping, indentation is used to keep track of scope.
For example, here the value of the `author` key is another mapping, setting `name` to the value `Soraya Drake`:

``` {.yaml filename="document.qmd"}
---
title: My document
author:
  name: Soraya Drake
---
```

It doesn't matter how many spaces you indent as long as you use spaces (as opposed to tabs), and use the same number consistently.

> A common style convention, and the one we follow in this book, is to use two spaces for each level of indentation.

Reviewer: ```suggestion
A common style convention, and the one we follow in this book and in Quarto documentation, is to use two spaces for each level of indentation.
```

For example, the `name` key nested under `author` is indented two spaces, whereas the `name` key nested under `affiliation`, which itself is nested under `author`, is indented four spaces:

``` {.yaml filename="document.qmd"}
---
title: My document
author:
  name: Soraya Drake
  affiliation: 
    name: Everpine University    
    url: http://everpine.edu
---
```

The names you can use in Quarto, along with the types of values they take, are not specified by the YAML language, but by Quarto's own schema.

### YAML Sequences

A YAML sequence is a way to provide a key a set of ordered values.
There are two syntaxes to specify a YAML sequence.
When the values in the sequence are simple, separate them with commas inside square brackets:

``` {.yaml filename="document.qmd"}
---
theme: [cosmo, styles.scss]
---
```

> The alternate syntax puts each value on their own line, and adds a dash prefix:

Reviewer: ```suggestion
The alternate syntax puts each value on its own line, and adds a dash prefix followed by a space:
```

``` {.yaml filename="document.qmd"}
---
theme: 
  - cosmo
  - styles.scss
---
```

This alternate syntax can also handle situations where the values are further mappings.
For example, here the value of `author` is a sequence of length two, where each element is a mapping of `name` and `affiliation`:

``` {.yaml filename="document.qmd"}
---
title: My document
author:
  - name: Soraya Drake
    affiliation: Everpine University    
  - name: Zahir Kamarov
    affiliation: Peregrine Heights Institute
---
```

### YAML Scalars

YAML uses the term *scalar* to talk about the simplest types of value: strings, numbers, and booleans.
You saw an example of each earlier:

``` {.yaml filename="document.qmd"}
---
title: My Document  # A string
toc: true           # A boolean, `true`/`false`
toc-depth: 2        # A number
---
```

> The only values recognized as boolean are `true` and `false`, numbers can be single integers (`2`), decimal (`2.1`) or scientific (`2.1e+2`), and anything else is treated as a string.

Reviewer: ```suggestion
The only values recognized as boolean are `true` and `false`; numbers can be single integers (`2`), decimals (`2.1`), or scientific (`2.1e+2`), and anything else is treated as a string.
```

### Quotes

Strings don't require quotes, but you may need them if your string contains characters that have special meaning in YAML (e.g. `#`, `:`, etc.), or is something that would otherwise be interpreted as a number or boolean:

``` {.yaml filename="document.qmd"}
---
title: "Quotes: and things that need them"  
author: "#quarto"
subtitle: "100"
---
```

> In YAML double quoted strings (`"`) allow escape characters in them, whereas single quoted strings (`'`) are interpreted literally:

Reviewer: ```suggestion
In YAML, double-quoted strings (`"`) allow escape characters in them, whereas single-quoted strings (`'`) are interpreted literally:
```

``` {.yaml filename="document.qmd"}
---
title: "Quotes: \U0001F4E8"
subtitle: 'Quotes: \U0001F4E8'
---
```

Q: where does the `\U` go in `subtitle` go?
YAML might be interpreting literally but Quarto/Pandoc might escape somewhere?

### Long strings

> For strings that are longer than you'd comfortably put on a single line, or need explicit line breaks, you can use what YAML calls block scalars.

Reviewer: What should we use to call out terms? Bolding? e.g., around **block scalars**.

Relevant but not just about this question -- are we doing an index?

There are two styles: literal and folded.

Literal block scalars are indicated by a pipe (`|`), and line breaks are considered significant, whereas folded block scalars are indicated by a greater than symbol (`>`), and line breaks are collapsed to spaces.

In both cases, start your string on a new line indented under the key:

::: {layout-ncol="2"}
``` {.yaml filename="literal.qmd"}
---
title: |
  A title

  spanning two lines
---
```

``` {.yaml filename="folded.qmd"}
---
title: >
  A title

  collapsed into one line
---
```
:::

::: {#fig-block-strings layout-ncol="2"}
![Rendered title using literal (`|`) block scalar](images/yaml-block-literal.png){#fig-block-strings-literal .border fig-alt="A screenshot of a title that reads 'A title' then a line break followed by 'spanning two lines'."}

![Rendered title using folded (`>`) block scalar](images/yaml-block-folded.png){#fig-block-strings-folded .border fig-alt="A screenshot of a title that reads 'A title collapsed into one line' all on a single line."}
:::

Depending on where the string value is used, line breaks may or may not mean anything to Quarto.

> For instance, when interpreted as markdown, a single line break doesn't usually result in anything in the output.

Reviewer: Is it usually or always?

Author: Always I think. Fixed.

However, a double line break will be interpreted as a new paragraph.

### Using `default`

You'll come across examples where the value of a Quarto option is `default`, like when you are specifying multiple formats:

``` {.yaml filename="document.qmd"}
---
format:
  html: default
  pdf: default
---
```

> The value `default` doesn't have a special YAML meaning--- it's something Quarto adds for certain options where the value is usually another mapping, but you just want the default values.

Reviewer: ```suggestion
The value `default` doesn't have a special YAML meaning--- it's something Quarto adds for certain options where the value is usually another mapping, but you just want the default values for all subfields.
```

Reviewer: I'm not sure if subfield is the right term here.

Author: Switched to "properties"

## What happens when you get it wrong?

> When you render a project or document, one the first things that happens is the options are parsed and validated.

Reviewer: ```suggestion
When you render a project or document, one of the first things that happens is the options are parsed and validated.
```

When you make a mistake in setting an option there are a few possible outcomes:

1.  You get an error message, and the error message clearly points out your mistake.

2.  You get an error message, but it's not clear from that message what or where your mistake is.

3.  You don't get an error message at all.

It's hard to enumerate all the ways you can make a mistake, but we'll give a couple of examples for each of these outcomes in the hope it helps you track down yours.

> Some examples in this chapter will result in an error from `quarto render`---we've highlighted these examples with a red border (TODO: what will we do in PDF), like this:

Reviewer: Could we do a red border and a red X icon/emoji or something for the div?

Author: Yeah, sounds good. I'll leave the TODO for now, because it's bound to need something special for LaTeX.

### Using the wrong YAML syntax

When your YAML can't be parsed by Quarto you'll get a `YAMLException` error.
These are actually surprisingly rare, not because everyone is great at YAML, but because most YAML syntax mistakes don't result in invalid YAML.
The YAML remains valid, it just won't be interpreted in the way you intended.

For example, if you forget the space after a colon when setting an option like `code-fold`:

``` {.yaml .error filename="document.qmd"}
---
title: "Untitled"
format:
  html:
    code-fold:true
---
```

Then you don't get a `YAMLException` but instead a validation error:

``` default
Validation of YAML front matter failed.
(line 5, columns 5--18) Field "html" has value code-fold:true, which must instead be an object

✖ The value code-fold:true is a string.
ℹ The error happened in location format:html.
ℹ In YAML, key-value pairs in objects must be separated by a space.
ℹ Did you mean code-fold: true instead?
```

The YAML has been parsed, but with the missing space `code-fold:true` was interpreted as a string not a mapping.
Quarto does a good job here suggesting what your mistake might have been and a solution.

> Generally, if you get a validation error, make a careful check of your syntax first.

Reviewer: ```suggestion
Generally, if you get a validation error, first check your syntax carefully.
```

### Specifying the wrong key

There are some instances where when you use the wrong name for an option you'll get an error.
For example, if you pluralize the `lightbox` option `loop`:

``` {.yaml .error filename="document.qmd"}
title: "Presentation"
format: html
lightbox: 
  loops: true 
```

You'll get a YAML validation error that catches and reports the problem:

``` default
ERROR: Validation of YAML front matter failed.
ERROR: In file long-string.qmd
(line 5, columns 3--8) property name loops is invalid
4: lightbox: 
5:   loops: true 
    ~~~~~~
ℹ The error happened in location lightbox:loops.
ℹ Did you mean loop?
```

> Options where Quarto only allows a certain set of property names are known as *closed schemas*.
> I.e. the `lightbox` option has a closed schema.

Reviewer: ```suggestion
Options where Quarto only allows a certain set of property names are known as *closed schemas*, e.g., the `lightbox` option has a closed schema.
```

Reviewer: Here a keyword/term is italicized. This goes back to my comment/question from earlier about how to mark these.

However, not all Quarto options have closed schemas.

When the schema is *open*, Quarto simply accepts an unknown option name and you get no error.
For example, you might use an underscore instead of a hyphen in `code-fold`:

``` {.yaml filename="document.qmd"}
---
title: My document
format: 
  html:
    code_fold: true
---
```

Even though `code_fold` isn't a recognized option for `html`, you'll get no error.

> But, you also won't get any cold folding.

Reviewer: ```suggestion
But, you also won't get any code folding.
```

Why does Quarto have open schemas?

> There's lots of places it makes sense to let users define their own keys, for example, when you want to use the `meta` shortcode and define some custom keys in your meta data.

Reviewer: ```suggestion
There are lots of places it makes sense to let users define their own keys, for example, when you want to use the `meta` shortcode and define some custom keys in your metadata.
```

A similar problem occurs when you nest options incorrectly.
For example, you might get `code-fold` right, but forget to nest `html` under `format`:

``` {.yaml filename="document.qmd"}
---
title: My document
html:
  code-fold: true
---
```

As a top-level option `html` isn't something Quarto knows about, but it passes validation.
The result is no error, and no folded code.

[Code completion](#yaml-completion) can be a useful indicator for these problems.
If you aren't getting the completion you expect, perhaps you've got the wrong name, or you've nested it under the wrong option.

### Specifying the wrong type of value

Checking the type of value you've passed is something Quarto is good at, assuming your syntax is correct and you've got the right option name.

For example, consider trying to add a table of contents like this:

``` {.yaml .error filename="document.qmd"}
---
title: My document
toc: yes
---
```

> On render Quarto returns the error:

Reviewer: ```suggestion
Upon render, Quarto returns the error:
```

Not all validation error messages are as illuminating.
For example, consider trying to add an "Other Link" like this:

``` {.yaml .error filename="document.qmd"}
---
title: "Untitled"
format:
  html:
    other-links: http://quarto.org
---
```

The validation error returned is:

``` default
ERROR: Validation of YAML front matter failed.
ERROR: In file type-errors.qmd
(line 5, columns 18--34) Field "other-links" has value http://quarto.org, which must instead be 'false'
4:   html:
5:     other-links: http://quarto.org
                   ~~~~~~~~~~~~~~~~~
6: ---
```

It's clear where the error is (`(line 5, columns 18--34) Field "other-links"`), and that the wrong type of value has been encountered.
But it's unclear what Quarto actually wants.
It says it was expecting `false`, but that doesn't help us figure out how to actually specify a link.

::: callout-tip
## Obtuse messages

Quarto tries not to overwhelm you with output in an error message, but sometimes the process of shortening an error message leads to a message that isn't very helpful.

This particular form of message that ends with `which must instead be x`, is a common culprit of being too brief to be helpful.

> Quarto will often report one value `x`, but in reality there are many values that the option could take.

Reviewer: Is the suggested `x` usually the first alphabetically among the options? Is that worth noting here?

:::

For validation errors like this you should look for an example in the Guide section of Quarto.
In the section ["Code and Other Links"](https://quarto.org/docs/output-formats/html-basics.html#code-links-and-other-links) you can see an example showing Quarto was expecting an array of objects:

``` {.yaml filename="document.qmd"}
---
title: "Untitled"
format:
  html:
    other-links: 
      - text: Quarto
        href: http://quarto.org
---
```

### Ways Quarto tries to help

> Quarto ships with YAML intelligence which encompasses two tools you can use to help you get your YAML right: YAML completion, and YAML validation.

Reviewer: ```suggestion
Quarto ships with YAML intelligence which encompasses two tools you can use to help you get your YAML right: YAML completion and YAML validation.
```

It is activated when you save your file.

It will warn you when:

-   a value has the incorrect type
-   a value isn't one of the enumerated allowable values
-   the name of an option isn't valid (but only in situations where free-form options aren't allowed)
-   indention is inconsistent

## What are the possible options?

The best place to find out what options are available is the [Reference section of the Quarto documentation](http://quarto.org/docs/reference).
These pages are automatically generated from the YAML schema defined in the source code of the Quarto CLI.

The options available depend on where the option will be applied.
From smallest to largest scope these are: to a code cell, to a document format, or to a project.

Those scopes also map to where you put the option: inside the code cell, in the document header, or in `_quarto.yml`.
However, you can generally apply options to a larger scope, by placing them in the corresponding location.
For example, `echo` is a code cell option, but you can use it in the document header to apply to formats, or you can use it in `_quarto.yml` to apply to all documents in a project.

> Table @tbl-options matches the sections of the Reference Guide, 

Reviewer: ```suggestion
@tbl-options matches the sections of the Reference Guide, 
```

|                         | Project metadata  |                             |
+-------------------------+-------------------+-----------------------------+
| Formats                 | Document options  | Document header             |
|                         |                   |                             |
|                         | Document metadata | `_quarto.yml`               |
|                         |                   |                             |
+-------------------------+-------------------+-----------------------------+
| Code Cells              | Execution options | Code cells, i.e. after `#|` |
|                         |                   |                             |
|                         |                   | Document header             |
|                         |                   |                             |
|                         |                   | `_quarto.yml`               |
+-------------------------+-------------------+-----------------------------+

: The three kinds of options you can set in Quarto {#tbl-options}

## Project options

Project options are literally those listed under `project` in `_quarto.yml`, 

> or under a project type (e.g. `website`, `book` or `manuscript`) in `_quarto.yml`.

Reviewer: ```suggestion
or under a project type (e.g., `website`, `book`, or `manuscript`) in `_quarto.yml`.
```
